OutOfSchool/OutOfSchool.WebApi/Services/TeacherService.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoMapper;
+using OutOfSchool.Services.Models;
+using OutOfSchool.Services.Repository;
+using OutOfSchool.WebApi.Mapping.Extensions;
+using OutOfSchool.WebApi.Models;
+
+namespace OutOfSchool.WebApi.Services
+{
+    /// <summary>
+    /// Service with business logic for Teacher model.
+    /// </summary>
+    public class TeacherService : ITeacherService
+    {
+        private IEntityRepository<Teacher> _repository;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TeacherService"/> class.
+        /// </summary>
+        /// <param name="repository">Repository for Teacher entity.</param>
+        public TeacherService(IEntityRepository<Teacher> repository)
+        {
+            _repository = repository;
+        }
+
+        /// <inheritdoc/>
+        public async Task<TeacherDTO> Create(TeacherDTO teacherDto)
+        {
+            if (teacherDto == null)
+            {
+                throw new ArgumentNullException($"{nameof(TeacherDTO)} entity must not be null");
+            }
+
+            try
+            {
+                var teacher = teacherDto.ToDomain();
+
+                var newTeacher = await _repository.Create(teacher).ConfigureAwait(false);
+
+                return newTeacher.ToModel();
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"{nameof(TeacherDTO)} could not be saved: {ex.Message}");
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task<IEnumerable<TeacherDTO>> GetAllTeachers()
+        {
+            try
+            {
+                var teachers = await _repository.GetAll().ConfigureAwait(false);
+                var teachersDto = teachers.Select(teacher => teacher.ToModel());
+
+                return teachersDto;
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Couldn't retrieve Teachers: {ex.Message}");
+            }
+        }
+
+        public async Task<TeacherDTO> GetById(long id)
+        {
+            var teacher = await _repository.GetById(id).ConfigureAwait(false);
+
+            if (teacher == null)
+            {
+                throw new ArgumentNullException($"There is no {nameof(teacher)} with id = {id}.");
```

Comment:
from msdn:
An ArgumentNullException exception is thrown when a method is invoked and at least one of the passed arguments is null but should never be null.

but this is not the case

Reply:
so as I understood, you propose to remove not needed checking for null at all?

Reply:
we need to use another exception type

Reply:
From my perspective, this type is the most suitable in this case. 
NullReferenceException is more general type, and it indicates nothing meaningful for us, as it could be caused by any number of reasons. 
ArgumentNullException is meaningful, because it tells us: the operation failed because the argument was null. We don't need to look at the code to figure that out.
That's what I just have read.
What is your suggestion?

Reply:
but an argument wasn't null
it's just that there is no such teacher
we might use ArgumentNullException or create our custom exception